sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1237,20 +1249,44 @@ class Analyzer(
       if (conflictPlans.isEmpty) {
         right
       } else {
-        val attributeRewrites = AttributeMap(conflictPlans.flatMap {
-          case (oldRelation, newRelation) => oldRelation.output.zip(newRelation.output)})
-        val conflictPlanMap = conflictPlans.toMap
-        // transformDown so that we can replace all the old Relations in one turn due to
-        // the reason that `conflictPlans` are also collected in pre-order.
-        right transformDown {
-          case r => conflictPlanMap.getOrElse(r, r)
-        } transformUp {
-          case other => other transformExpressions {
+        rewritePlan(right, conflictPlans.toMap)._1
+      }
+    }
+
+    private def rewritePlan(plan: LogicalPlan, conflictPlanMap: Map[LogicalPlan, LogicalPlan])
+      : (LogicalPlan, mutable.ArrayBuffer[(Attribute, Attribute)]) = {
```

Comment:
nit: In the most cases of recursive calls, a return type would be `Seq` instead of `ArrayBuffer`?

Reply:
We can return `Seq` here. But I think it's not needed if the  `attrMapping` below is still an `Arraybuffer`. Or, you mean the `attrMapping` could also be `Seq`? (Please also note the `attrMapping` has to be an mutable collection anyway.)

Reply:
Yea, cosmetic changes though, I thought it like this;
```
@@ -1254,16 +1254,15 @@ class Analyzer(
     }
 
     private def rewritePlan(plan: LogicalPlan, conflictPlanMap: Map[LogicalPlan, LogicalPlan])
-      : (LogicalPlan, mutable.ArrayBuffer[(Attribute, Attribute)]) = {
-      val attrMapping = new mutable.ArrayBuffer[(Attribute, Attribute)]()
+      : (LogicalPlan, Seq[(Attribute, Attribute)]) = {
       if (conflictPlanMap.contains(plan)) {
         // If the plan is the one that conflict the with left one, we'd
         // just replace it with the new plan and collect the rewrite
         // attributes for the parent node.
         val newRelation = conflictPlanMap(plan)
-        attrMapping ++= plan.output.zip(newRelation.output)
-        newRelation -> attrMapping
+        newRelation -> plan.output.zip(newRelation.output)
       } else {
+        val attrMapping = new mutable.ArrayBuffer[(Attribute, Attribute)]()
         var newPlan = plan.mapChildren { child =>
           // If not, we'd rewrite child plan recursively until we find the
           // conflict node or reach the leaf node.
```

Reply:
I see. I can follow your suggestion. Thanks!